### PR TITLE
Issue-46: log action summary after syncing is done

### PR DESCRIPTION
**Pull request by ensembledeanw**
PR for Issue #46: add action summary to log

### CODE REVIEW

File: user_sync/rules.py [hunk 1]
```diff
@@ -64,6 +64,19 @@ def __init__(self, caller_options):
...
+        self.adobe_users = set()
+        # counters for action summary log
+        self.action_summary = {
+            'number_of_adobe_users': 0,
+            'number_of_directory_users': 0,
+            'number_of_users_created': 0,
+            'number_of_users_updated': 0,
+            'number_of_users_with_updated_groups': 0,
+            'number_of_users_removed_from_mapped_group': 0,
+            'number_of_users_remove_or_deleted': 0,
```
Comment (adobeDan) [inline user_sync/rules.py]: removed_or_deleted?  (you have remove_or_deleted)

File: user_sync/rules.py [hunk 2]
```diff
@@ -140,7 +153,26 @@ def run(self, directory_groups, directory_connector, dashboard_connectors):
...
             
+    def log_action_summary(self):
+        '''
+        log number of directory and Adobe users, number of created / updated / deleted users, and number of users with updated groups
+        or removed from mapped group because they were not in directory
+        :return: None
+        '''
+        logger = self.logger
+        # find the total number of directory users
+        self.action_summary['number_of_directory_users'] = len(self.filtered_directory_user_by_user_key);
```
Comment (adobeDan) [inline user_sync/rules.py]: You need to report both the total number of directory users and the number of directory users in the "selected" input set (i.e., filtered set).

File: user_sync/rules.py [hunk 3]
```diff
@@ -140,7 +153,26 @@ def run(self, directory_groups, directory_connector, dashboard_connectors):
...
+        # find the total number of adobe users
+        self.action_summary['number_of_adobe_users'] = len(self.adobe_users);
```
Comment (adobeDan) [inline user_sync/rules.py]: You need to report both the total number of Adobe users and the number that were "excluded from operations" (via the exclude_parameters).  Look at the `self.excluded_users_keys` variable for this count.

File: user_sync/rules.py (hunk 1 above)
Comment (ensembledeanw) [inline user_sync/rules.py]: I will separate the counter. One for removed and one for deleted

**Comment by ensembledeanw** [comment]
Hi Dan, I separated the count for deleted user and removed user. And I also added counters for excluded Adobe users and selected directory users. It is ready for you to review again. Thanks

**Comment by ensembledeanw** [comment]
Hi Dan, Sorry I didn't see your request for adding English description. Let me do it now.

**Comment by ensembledeanw** [comment]
Hi Dan, I added the English text description and print the description instead of variable name. It is ready for review again. Thanks

File: user_sync/rules.py [hunk 4]
```diff
@@ -75,11 +75,26 @@ def __init__(self, caller_options):
...
             'number_of_users_with_updated_groups': 0,
-            'number_of_users_removed_from_mapped_group': 0,
+            'number_of_users_removed_from_mapped_groups': 0,
             'number_of_users_removed': 0,
             'number_of_users_deleted': 0,
             'number_of_users_with_no_changes': 0,
         }
+        # English text description for action summary log.
+        # The action summary will be shown the same order as they are defined in the list
+        self.action_summary_descriptions = [
```
Comment (adobeDan) [inline user_sync/rules.py]: Probably better to make this a file-level or class-level constant.

File: user_sync/rules.py [hunk 5]
```diff
@@ -75,11 +75,26 @@ def __init__(self, caller_options):
...
+            ['total_number_of_adobe_users', 'Total number of Adobe users'],
+            ['number_of_adobe_users_excluded', 'Number of Adobe users excluded'],
+            ['total_number_of_directory_users', 'Total number of directory users'],
+            ['number_of_directory_users_selected', 'Number of directory users selected'],
+            ['number_of_users_created', 'Number of users created'],
```
Comment (adobeDan) [inline user_sync/rules.py]: let's be clear in all these "change" statistics that they are "Adobe users", not just "users"

**Comment by ensembledeanw** [comment]
Hi Dan, I have made changes to the code base on your comments. It is ready for review.

**Comment by ensembledeanw** [comment]
Hi Dan, I just see your email about the document. I will do that after dinner.

**Comment by ensembledeanw** [comment]
Hi Dan, I added the section about action summary into user manual. Thanks

**Pull request merged**